perf: optimize compute_block_hash_for_seq with unsafe pointer casting

xiaguan opened the pull request:
## Summary
This PR optimizes the `compute_block_hash_for_seq` function in the KV router indexer by eliminating memory allocations during hash computation. The optimization uses unsafe pointer casting to reinterpret `u32` slices as `u8` slices without copying data, achieving **66-76% performance improvement** across various workloads.

## Performance Results

| Test Case | Original Time | Optimized Time | Improvement |
|-----------|---------------|----------------|-------------|
| len_10000_chunk_16 | 17.037 µs | 3.127 µs | **81.6% faster** |
| len_10000_chunk_64 | 7.758 µs | 4.140 µs | **46.6% faster** |
| len_50000_chunk_16 | 85.231 µs | 15.325 µs | **82.0% faster** |
| len_50000_chunk_64 | 38.464 µs | 20.617 µs | **46.4% faster** |

## Technical Details

### Before (Original Implementation)
```rust
pub fn compute_block_hash_for_seq(tokens: &[u32], kv_block_size: u32) -> Vec<LocalBlockHash> {
    tokens
        .chunks_exact(kv_block_size as usize)
        .map(|chunk| {
            let bytes: Vec<u8> = chunk
                .iter()
                .flat_map(|&num| num.to_le_bytes())
                .collect();
            compute_block_hash(&Bytes::from(bytes))
        })
        .collect()
}
```

**Performance bottlenecks:**
- Memory allocation for each `Vec<u8>` per chunk
- Iterator overhead with `flat_map` and `collect()`
- Additional allocation when converting `Vec<u8>` to `Bytes`

### After (Optimized Implementation)
```rust
pub fn compute_block_hash_for_seq(tokens: &[u32], kv_block_size: u32) -> Vec<LocalBlockHash> {
    tokens
        .chunks_exact(kv_block_size as usize)
        .map(|chunk| {
            // SAFETY: This is safe because:
            // 1. u32 and u8 have compatible memory layouts (u32 is 4 contiguous u8s)
            // 2. The slice is valid for the duration of this operation
            // 3. We're only reading from the memory, not modifying it
            // 4. The alignment requirements are satisfied (u8 has no alignment requirements)
            let bytes = unsafe {
                std::slice::from_raw_parts(
                    chunk.as_ptr() as *const u8,
                    chunk.len() * std::mem::size_of::<u32>(),
                )
            };
            compute_block_hash(bytes)
        })
        .collect()
}
```

**Optimization benefits:**
- **Zero memory allocations** during byte conversion
- **Direct memory reinterpretation** instead of copying
- **Maintains identical behavior** and function signature


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

* **Tests**
  * Added a new test to verify the correctness of block hash computation for various token sequences and block sizes.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

xiaguan committed:
perf: optimize compute_block_hash_for_seq with unsafe pointer casting

- Replace Vec<u8> allocation with unsafe slice reinterpretation
- Use std::slice::from_raw_parts to avoid memory copying
- Achieve 66-76% performance improvement in hash computation
- Add comprehensive test to verify correctness against reference implementation
- Maintain same function signature and behavior
- Add detailed safety comments for unsafe code usage

The optimization eliminates memory allocations when converting u32 chunks
to bytes for hashing by directly reinterpreting the memory layout instead
of copying data through iterators and collections.

xiaguan committed:
Fix CI suggestions: remove unused import and use std::mem::size_of_val

- Remove unused bytes::Bytes import
- Replace manual slice size calculation with std::mem::size_of_val(chunk)
- Addresses clippy::manual-slice-size-calculation warning

ryanolson commented:
Nice. Good spot. I think we can use bytemuck to make this even cleaner.

ryanolson commented on lib/llm/src/kv_router/indexer.rs:
```diff
@@ -123,12 +122,19 @@ pub fn compute_block_hash_for_seq(tokens: &[u32], kv_block_size: u32) -> Vec<Loc
     tokens
         .chunks_exact(kv_block_size as usize) // Split into chunks of kv_block_size elements
         .map(|chunk| {
```
.map(|chunk| compute_block_hash(bytemuck::cast_slice::<u32, u8>(chunk)))

ryanolson replied:
But I also wonder if we need to worry about the router running on a different arch than the worker. 

If one was amd64 and the other arm64, we might need to map to LE (or BE) for consistency.

ryanolson replied:
We also have in token.rs which is using both rayon (par_chunk) and cast_slice.

xiaguan committed:
refactor(indexer): ensure consistent hashing across architectures

xiaguan commented:
Thanks a lot for your thorough review and suggestions!

I've replaced the unsafe code with `bytemuck`, and added a fallback for big-endian platforms using`cfg`.

I also gave `rayon` a try, but the tasks were too small to really benefit from it—in fact, the overhead ended up making things slower. So I decided to not use it.
